ktor-network/jvm/test/io/ktor/network/sockets/tests/ClientSocketTest.kt

```diff
@@ -95,45 +99,35 @@ class ClientSocketTest {
     @Test
     fun testSelfConnect() {
         runBlocking {
-            // Find a port that would be used as a local address.
-            val port = getAvailablePort()
-
-            val tcpSocketBuilder = aSocket(ActorSelectorManager(Dispatchers.IO)).tcp()
-            // Try to connect to that address repeatedly.
-            for (i in 0 until 100000) {
-                try {
-                    val socket = tcpSocketBuilder.connect(InetSocketAddress("127.0.0.1", port))
-                    fail("connect to self succeed: ${socket.localAddress} to ${socket.remoteAddress}")
-                } catch (ex: Exception) {
-                    // ignore
+            selector.buildOrClose({ openSocketChannel() }) {
+                val clientAddress = InetSocketAddress("0.0.0.0", 3333)
+                val clientSocket = object: java.net.Socket() {
+                    override fun getRemoteSocketAddress(): SocketAddress = clientAddress
+                    override fun getLocalSocketAddress(): SocketAddress = clientAddress
+
+                    override fun close() {
+                        super.close()
+                        server?.first?.close()
+                        this@buildOrClose.close()
+                    }
                 }
-            }
-        }
-    }
-
-    // since new linux kernel version introduce a feature, new bind port number will always be odd number
-    // and connect port will always be even, so we find a random even port with while loop
-    // see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=07f4c90062f8fc7c8c26f8f95324cbe8fa3145a5
-    private fun getAvailablePort(): Int {
-        while (true) {
-            val port = ServerSocket().apply {
-                bind(InetSocketAddress("127.0.0.1", 0))
-                close()
-            }.localPort
-
-            if (port % 2 == 0) {
-                return port
-            }
 
-            try {
-                // try bind the next even port
-                ServerSocket().apply {
-                    bind(InetSocketAddress("127.0.0.1", port + 1))
-                    close()
+                try {
+                    server { }
+                    nonBlocking()
+                    SocketImpl(
+                        this,
+                        clientSocket,
+                        ActorSelectorManager(Dispatchers.IO)
```

Comment:
This should be closed explicitly. I know we had it before, but would be nice to fix once you are changing the test